Add support for 404 instead of dir listing

ethanaobrien opened the pull request:
Add support for 404 instead of dir listing (#262 )

Tested by me.
I changed the version in the manifest to 2.5.3, you can change it if you like. 

It was really hard at first, but it got easier once I figured out what was where.

I also just replaced the broken image in the readme with the default logo

kzahel commented:
Hi @ethanaobrien thanks so much for all the activity on this project. I'll take a look at your branch in a bit.
I haven't been investing much because of chrome apps being sunsetted within a year or so. 
Support for osx/windows has already ended (though it keeps working). I would expect that by the end of the year some chrome update will just disable apps (and make this no longer work).
Still, I think you're doing some great changes here.

kzahel commented:
Also keep in mind having a huge pull request that is not specific to any single issue makes it much harder to evaluate the changes. I would definitely prefer smaller pull requests in the future.

ethanaobrien commented:
@kzahel 
Ok, Thank you! I will do that in the future. I plan to add a feature like htaccess (Not exactly, just does the same things). This will take a bit more time (as I will need to do a lot of change).

I did catch a bug for users not on chrome os with the https setting auto restarting.
With the custom 404 page, I made sure every file (Like .js files) render as an html and not code.

Sorry for the mess! I will do multiple smaller ones in the future!

ethanaobrien commented:
@kzahel I added support for this wsc.htaccess feature that I have been working on. The reason why the commit labeled "add wsc.htaccess support" does not say Verified next to it is because I used git to push my changes

ethanaobrien commented:
@kzahel one of my comments got berried by the commits. Click the view more to see it. Sorry about the mess

ethanaobrien commented:
@kzahel Think about the size and date modified in directory listing as like a payback for the huge mess. I hope you like the features I added :)

ethanaobrien commented:
@kzahel I have finished this. Every feature is in and I have gone looking and squashed all of the bugs.

I will not let the amount of commits get out of control in the future. For this I am sorry.

kzahel commented:
Hi @ethanaobrien . I haven't had a chance to check out your branch and test it yet. I see you are still making a few final commits. I get a lot of email notifications from you now. If something is more urgent, feel free to @ me. I will try to make some time this weekend to take a look at all your changes. Thanks again for all the comments in the issues, I will also try and make a pass and close some issues and do some updating as you pointed out in those various issues.

ethanaobrien commented:
Awesome! But yeah. I've decided not to make any more commits until this pr gets looked at so that is what I am going to wait for.

Thank you

ethanaobrien commented:
Also, when (and if) you merge this, before you put it on the chrome web store there is a link to the htaccess readme that you will need to put into the `react-ui/js/index.js` folder (There is also a link in the readme)

kzahel commented on webapp.js:
```diff
@@ -495,9 +496,56 @@
+                    file.file( function(filee) {
+                    var reader = new FileReader();
+                    reader.onload = function(e){
+                        this.useDefaultMime = false
```
This is pretty weird to switch on and off a global variable like this. If I was still actively maintaining this project I would ask you to do something less weird here :-)

ethanaobrien replied:
Yeah I think the same thing. It was the easiest way I could think of the mime type not rendering as the requested file extension.

kzahel commented on makedeps/package.json:
```diff
@@ -12,8 +12,8 @@
-    "react": "^16.12.0",
-    "react-dom": "^16.12.0",
+    "react": "^17.0.2",
+    "react-dom": "^17.0.2",
```
cool looks like the upgrade worked

kzahel commented on directory-listing-template.html:
```diff
@@ -1,52 +1,48 @@
 <!DOCTYPE html>
 
-<html i18n-values="dir:textdirection;lang:language">
+<html dir="ltr" lang="en">
```
oh nice. i copied this from chrome's builtin file:/// protocol viewer, i guess i didn't notice this

ethanaobrien replied:
I copied this from chromes built in thing too, I just used an updated version

kzahel commented on handlers.js:
```diff
@@ -94,17 +94,167 @@
+                                    var filefound = true
+                                    break
+                                }
+                            }
```
it looks like there's a lot of code duplication between delete and put access checks. It could probably be simplified if you wanted to (I don't care)

ethanaobrien replied:
I thought about doing that, but It would take a lot of works. It works perfectly at the moment so I dont plan on changing it.

kzahel commented on handlers.js:
```diff
@@ -159,6 +309,31 @@
+            if (this.app.opts.optExcludeDotHtml && ! this.request.origpath.endsWith("/")) {
+                var extension = this.request.path.split('.').pop();
+                var more = this.request.uri.split('.html').pop()
+                if (extension == 'html') {
```
people used to also write `.htm` files though I guess that's back from a long time ago :-)

ethanaobrien replied:
Yeah I did think about that. If someone makes an issue I may implement it but that would make the whole thing just crazy

kzahel commented on handlers.js:
```diff
@@ -318,14 +508,330 @@
                 reader.readEntries( onreadsuccess.bind(this),
                                     onreaderr.bind(this))
             }
+            }
```
weird indentation

ethanaobrien replied:
Yeah, sorry about that. The syntax got a little messy and not everything was lined up

kzahel commented on handlers.js:
```diff
@@ -318,14 +508,330 @@
+                                                                this.useDefaultMime = false
+                                                                var data = e.target.result
+                                                                this.setHeader('content-type','text/html; charset=utf-8')
+                                                                this.setHeader("WWW-Authenticate", "Basic")
```
Cool I didn't know you could add basic auth in htaccess.
It might be also a good idea to just have a global basic auth option in the settings (without having to learn the htaccess stuff). e.g. "Use basic auth" in advanced settings. Very cool!

ethanaobrien replied:
Yeah! I was doing some digging in the webapp.js and I found that this was already implemented! We just need to do the UI. (I am not marking this as resolved. I am going to do this real quick)

kzahel commented on handlers.js:
```diff
@@ -318,14 +508,330 @@
+                                            filerequested == 'index.htm' ||
+                                            filerequested == 'index' ||
+                                            filerequested == 'index.xhtm' ||
+                                            filerequested == 'index.xhtml' ||
```
this would be much cleaner to make a helper util function, instead of duplicating it in several places

ethanaobrien replied:
Thats a good Idea. I may do that but its not likely.

kzahel commented on handlers.js:
```diff
@@ -415,28 +916,132 @@
+                            //console.log(file)
+                            var rawname = results[w].name
+                            //from https://stackoverflow.com/questions/10420352/converting-file-size-in-bytes-to-human-readable-string/10420404
+                            function humanFileSize(bytes, si=false, dp=1) {
```
for better organization this would be better defined in like a utils.js file or something

ethanaobrien replied:
I think the same thing, I tried to do it and everything got a little crazy which is why I just put it here.

kzahel commented on handlers.js:
```diff
@@ -415,28 +916,132 @@
+                            var lastModifiedYear = file.modificationTime.getFullYear().toString().substring(2, 4)
+                            var lastModifiedHours = file.modificationTime.getHours() + 1
+                            var lastModifiedMinutes = file.modificationTime.getMinutes() + 1
+                            var lastModifiedSeconds = file.modificationTime.getSeconds() + 1
```
are you sure you add 1 to seconds/minutes? I think you only have to do that for month? looks weird..

ethanaobrien replied:
I'll look at this

ethanaobrien replied:
Fixed

kzahel commented on handlers.js:
```diff
@@ -415,28 +916,132 @@
+                                                  lastModifiedHours+':'+
+                                                  lastModifiedMinutes+':'+
+                                                  lastModifiedSeconds +' '+
+                                                  lastModifiedAmPm
```
maybe better to just use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat or something

ethanaobrien replied:
Thank you! I will do this

kzahel commented on handlers.js:
```diff
@@ -415,28 +916,132 @@
+                        var html = ['<script>start("'+this.request.path+'")</script>',
+                                    '<script>onHasParentDirectory();</script>']
+                    } else {
+                        var html = ['<script>start("/")</script>']
```
i don't see how these html are being used. (you'd need to writeChunk them?)

ethanaobrien replied:
Sorry about that. I just did what I knew and it sometimes gets a little messy. Sorry about that

kzahel commented on react-ui/js/index.js:
```diff
@@ -102,7 +122,6 @@ const functions = {
     app.webapp.updateOption('optUseHttps', val);
     if (app.webapp.started) {
       app.webapp.stop();
-      app.webapp.start();
```
maybe it could auto start if they already generated the certificates. but yeah maybe this is fine

ethanaobrien replied:
I don't think that is the problem. I have tried changing to many different methods of restarting and it simply isn't possible. The reason it broke was it was trying to start the server while it was stopping (javascript synchronous functions). The only really way is to use like a delay function in javascript (and I know none)

kzahel commented on webapp.js:
```diff
@@ -316,7 +316,8 @@
 					var extIP = gateway.device.externalIP
-					this.extra_urls = [{url:'http://'+extIP+':' + this.port}]
+					let prot = this.opts.optUseHttps ? 'https' : 'http';
+					this.extra_urls = [{url:prot+'://'+extIP+':' + this.port}]
```
nice fix

kzahel reviewed (approved):
Looks good overall! Great work. I left some feedback but nothing that I need you to change, I'm ok to merge as-is.

kzahel commented:
I'll wait to merge until you want me to; in case you want to make any final changes in this PR. I'll plan on doing a release after merging and testing a little bit more.

ethanaobrien commented:
@kzahel Everything has been finished and it is ready to merge!

Also, I saw all the comments you left on the other issues and I want to say thank you!

ethanaobrien commented:
@kzahel In my latest 2 commits, I have added a htaccess feature to send additional headers and I have re done all of the for() statements that are scanning the htaccess file. I cleaned up handlers.js by making a utils.js file. I also added another htaccess feature to send the current directory with the requested file.

ethanaobrien commented:
@kzahel Please don't forget about this pr. Also, I found a way (In my latest commit) to restart the web server!

ethanaobrien commented:
Accidentally marked commit 26cf1c3 as Log to file - Meant to say Added folder creation for log to file and replaced all tabs with groups of 4 spaces

ethanaobrien commented:
Just so you know, this will always be ready to merge. I'm going to keep doing things as they come to mind until this is merged.

kzahel commented:
I see you keep committing :-)
I liked the changes you made when I tested it last but I haven't had the time to look at your most recent changes.

I will do my best to actually test it again, merge, and make a release this coming weekend (no promises though)

ethanaobrien commented:
@kzahel Thank you, Below I am going to provide some info on the ip feature

The ip feature will only show the users ip if not on a local address (you need to use port forwarding)
I opened my server yesterday to run some tests (and opened port forwarding) and, using the log to file feature, I was able to catch some spam requests. One is listed below.

```
clearIdle

209.141.47.35: Request GET /config/getuser?index=0 

36 response code 404 clen 1260 

http conn close
```

I just think it is funny when I see requests like these (Because I've tested and caught every way possible to do anything even remotely like this)

I also used this method to test my ip blocking feature.

Thank you! This is a really fun project!

ethanaobrien commented:
@kzahel Because of [this](https://github.com/terreng/web-server/pull/2). I will no longer be working on this. You can merge this when you are ready!

dpm2682 reviewed (approved):
Ok

ethanaobrien commented:
@kzahel Do you think you could look into merging this?

ethanaobrien commented:
Marked as draft since I don't entirely think this should be merged
